rbx_types/src/terrain/smooth_grid.rs

```diff
@@ -0,0 +1,670 @@
+                        voxel_count % CHUNK_SIZE,
+                    ),
+                    Voxel {
+                        material,
+                        solid_occupancy: occupancy.unwrap_or(0.0),
+                        water_occupancy: water_occupancy.unwrap_or(0.0),
+                    },
+                );
+                voxel_count += 1;
+            }
+        }
+
+        Ok(chunk)
+    }
+}
+
+/// A container allowing the modification, encoding, and decoding of the
+/// `SmoothGrid` data used by Roblox's `Terrain` object.
+#[derive(Debug, Clone, Default, PartialEq)]
+#[cfg_attr(
+    feature = "serde",
+    derive(serde::Serialize, serde::Deserialize),
+    serde(transparent)
+)]
+pub struct SmoothGrid {
+    world: BTreeMap<ChunkCoordinates, Chunk>,
+}
+
+impl SmoothGrid {
+    /// Constructs a new `Terrain` with no chunks.
+    #[inline]
+    pub fn new() -> Self {
+        Self {
+            world: BTreeMap::new(),
+        }
+    }
+
+    /// Finds a `Chunk` at the given position in this `Terrain`,
+    /// returning a reference to it if it exists.
+    #[inline]
+    pub fn get_chunk(&self, position: &ChunkCoordinates) -> Option<&Chunk> {
+        self.world.get(position)
+    }
+
+    /// Finds a `Chunk` at the given position in this `Terrain`,
+    /// returning a mutable reference to it if it exists.
+    #[inline]
+    pub fn get_chunk_mut(&mut self, position: &ChunkCoordinates) -> Option<&mut Chunk> {
+        self.world.get_mut(position)
+    }
+
+    /// Writes (or overwrites) a `Chunk` at the given position in this `Terrain`.
+    #[inline]
+    pub fn write_chunk(&mut self, position: &ChunkCoordinates, chunk: Chunk) {
+        self.world.insert(*position, chunk);
+    }
+
+    pub fn encode(&self) -> Vec<u8> {
+        let mut data = Vec::with_capacity(self.world.len() * 512);
+        data.extend([0x01, CHUNK_SIZE.ilog2() as u8]);
+
+        let mut chunk_cursor = None;
+        for (position, chunk) in &self.world {
+            let cursor = chunk_cursor.unwrap_or(position);
+            let axes = [
+                position.0.x - cursor.0.x,
+                position.0.y - cursor.0.y,
+                position.0.z - cursor.0.z,
+            ];
+
+            write_interleaved_i32_array(&mut data, axes.iter().copied()).unwrap();
+
+            data.extend(chunk.encode());
+            chunk_cursor = Some(position);
+        }
+
+        data
+    }
+
+    /// Decodes a `SmoothGrid` from a binary blob. The blob must be the same format used
+    /// by `encode` and Roblox.
+    pub fn decode(mut buffer: &[u8]) -> Result<Self, CrateError> {
+        let mut header = [0u8, 0u8];
+        if let Err(e) = buffer.read_exact(&mut header) {
+            return Err(SmoothGridError::from(e).into());
+        }
+
+        let [version, chunk_size] = header;
+        if version != 0x01 || chunk_size != 0x05 {
+            return Err(SmoothGridError::InvalidHeader(version, chunk_size).into());
+        }
```

Comment:
We should move these two be constants. Presumably then `CHUNK_SIZE` can be derived from the size we expect/support.

Then again, is it going to kill us to support a chunk size other than `0x05`? What would the consequence be?

Reply:
When the engine loads chunk sizes other than `0x05`, if I recall according to my testing, it just converts the data into a normal chunk size. The behavior is pretty untested and I'm not sure if there's hidden issues with a different chunk size.

Reply:
I suppose that's fair. We should probably just use Roblox's default.

Still, we should move both the version and chunk size to constants so that they're not just magic numbers floating around.